saleor/dashboard/product/forms.py

```diff
@@ -2,77 +2,161 @@
+        fields = []
+
+    def __init__(self, *args, **kwargs):
+        super(VariantAttributesForm, self).__init__(*args, **kwargs)
+        self.available_attrs = self.instance.product.attributes.prefetch_related('values').all()
+        for attr in self.available_attrs:
+            field_defaults = {'label': attr.display, 'required': True,
+                              'initial': self.instance.get_attribute(attr.pk)}
+            if attr.has_values():
+                choices = [('', '')] + [(value.pk, value.display)
+                                        for value in attr.values.all()]
+                field = forms.ChoiceField(choices=choices, **field_defaults)
+            else:
+                field = forms.CharField(**field_defaults)
+            self.fields[attr.get_formfield_name()] = field
+
+    def save(self, commit=True):
+        attributes = {attr.pk: self.cleaned_data.pop(attr.get_formfield_name())
+                      for attr in self.available_attrs}
+        self.instance.attributes = attributes
+        return super(VariantAttributesForm, self).save(commit=commit)
 
 
 def get_product_form(product):
-    if isinstance(product, Shirt):
-        return ShirtForm
-    elif isinstance(product, Bag):
-        return BagForm
+    if isinstance(product, Product):
+        return ProductForm
     else:
-        raise ValueError('Unknown product')
+        raise ValueError('Unknown product class')
 
 
 def get_product_cls_by_name(cls_name):
-    if not cls_name in PRODUCT_CLASSES:
+    for cls in PRODUCT_CLASSES.keys():
+        if cls_name == cls.__name__:
+            return cls
+    raise ValueError('Unknown product class')
+
+
+def get_variant_form(product):
+    if isinstance(product, Product):
+        return ProductVariantForm
+    else:
         raise ValueError('Unknown product class')
-    return PRODUCT_CLASSES[cls_name]
 
 
-def get_variant_formset(product):
-    if isinstance(product, Shirt):
-        return ShirtVariantFormset
-    elif isinstance(product, Bag):
-        return BagVariantFormset
+def get_variant_cls(product):
+    if isinstance(product, Product):
+        return ProductVariant
     else:
-        raise ValueError('Unknown product')
+        raise ValueError('Unknown product class')
+
+
+class ProductImageForm(forms.ModelForm):
+    class Meta:
+        model = ProductImage
+        exclude = ('product', 'order')
+
+
+class ProductAttributeForm(forms.ModelForm):
+    class Meta:
+        model = ProductAttribute
+        exclude = ['slug']
+
+    def __init__(self, *args, **kwargs):
+        super(ProductAttributeForm, self).__init__(*args, **kwargs)
+
+    def clean(self):
+        cleaned_data = super(ProductAttributeForm, self).clean()
+        slug = slugify(self.cleaned_data['display'])
+        if ProductAttribute.objects.filter(slug=slug):
+            raise forms.ValidationError(
+                pgettext_lazy('Validation error', 'Attribute already exists'))
```

Comment:
You use an automatically generated field to enforce uniqueness but the user has no way to correct said field if two different display names result in the same slug being generated.